src/libraries/Microsoft.Extensions.Logging/gen/LoggerMessageGenerator.Parser.cs

```diff
@@ -0,0 +1,659 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.Extensions.Logging.Generators
+{
+    public partial class LoggerMessageGenerator
+    {
+        internal class Parser
+        {
+            private readonly CancellationToken _cancellationToken;
+            private readonly Compilation _compilation;
+            private readonly Action<Diagnostic> _reportDiagnostic;
+
+            public Parser(Compilation compilation, Action<Diagnostic> reportDiagnostic, CancellationToken cancellationToken)
+            {
+                _compilation = compilation;
+                _cancellationToken = cancellationToken;
+                _reportDiagnostic = reportDiagnostic;
+            }
+
+            /// <summary>
+            /// Gets the set of logging classes containing methods to output.
+            /// </summary>
+            public IReadOnlyList<LoggerClass> GetLogClasses(IEnumerable<ClassDeclarationSyntax> classes)
+            {
+                const string LoggerMessageAttribute = "Microsoft.Extensions.Logging.LoggerMessageAttribute";
+
+                INamedTypeSymbol loggerMessageAttribute = _compilation.GetTypeByMetadataName(LoggerMessageAttribute);
```

Comment:
Is this the "100% safe" way to look up these types? I know Jared mentioned a few different approaches for how to find types the correct way in a source generator.

Reply:
Confirmed with @chsienki offline, and seems like GetTypeByMetadataName is the right approach, since the compiler will handle all the aliasing, full name. etc.

Reply:
The `GetTypeByMetadataName` API will return `null` when there are multiple types with the same FQN in the `Compilation`. That may or may not be a concern for you here. Basically do you want the logger to still work if a customer defines a type with that name in their library? 

If not then there are essentially two approaches you can use:

- If you care about using the type defined in a specific assembly then navigate to the `IAssemblySymbol` where that type is defined and use `GetTypeByMetadataName` on it. A FQN will be unique within an assembly. 
- If you don't care about the assembly, you just want to run whenever an attribute with that FQN is used, then bind the attribute, grab it's type, print out it's full name and see if it matches `Microsoft.Extensions.Loggnig.LoggerMessageAttribute`.